File: 0014/merge_k_sorted_linked_lists.py
# https://neetcode.io/problems/merge-k-sorted-linked-lists

from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def merge_k_lists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists or len(lists) < 1: return None

        while len(lists) > 1:
            merged_lists = []

            for i in range(0, len(lists), 2):
                first_list = lists[i]
                second_list = lists[i + 1] if i + 1 < len(lists) else None

                merged_lists.append(self.merge_lists(first_list, second_list))

            lists = merged_lists

        return merged_lists[0]

    def merge_lists(self, first_list, second_list):

        result = ListNode()
        tail = result

        while first_list and second_list:
            if first_list.val < second_list.val:
                tail.next = first_list
                first_list = first_list.next
            else:
                tail.next = second_list
                second_list = second_list.next

            tail = tail.next

        if first_list: tail.next = first_list
        if second_list: tail.next = second_list

        return result.next

File: 0014/merge_k_sorted_linked_lists.dart
void main() {
  final lists = [
    ListNode.fromArray([1, 2, 4]),
    ListNode.fromArray([1, 3, 5]),
    ListNode.fromArray([3, 6]),
  ];

  final result = mergeKLists(lists);
  print(result?.toList());
}

class ListNode<T> {
  ListNode({required this.val, this.next});

  final T val;
  ListNode<T>? next;

  factory ListNode.fromArray(List<T> array) {
    if (array.isEmpty) throw Exception('Array cannot be empty');
    final head = ListNode(val: array.first);
    ListNode<T>? tail = head;

    for (int i = 1; i < array.length; i++) {
      tail?.next = ListNode(val: array[i]);
      tail = tail?.next;
    }
    return head;
  }

  List<T> toList() {
    final result = [val];

    var curr = next;
    while (curr != null) {
      result.add(curr.val);
      curr = curr.next;
    }
    return result;
  }

  @override
  String toString() => toList().toString();
}

ListNode<int>? mergeKLists(List<ListNode<int>>? nodes) {
  if (nodes == null || nodes.isEmpty) return null;

  while (nodes!.length > 1) {
    final mergedLists = <ListNode<int>>[];

    for (int i = 0; i < nodes.length; i += 2) {
      final firstNode = nodes[i];
      final secondNode = i + 1 < nodes.length ? nodes[i + 1] : null;

      mergedLists.add(mergeLists(firstNode, secondNode));
    }

    nodes = mergedLists;
  }

  return nodes.first;
}

ListNode<int> mergeLists(ListNode<int>? firstNode, ListNode<int>? secondNode) {
  final dummy = ListNode(val: 0);
  ListNode<int>? tail = dummy;

  while (firstNode != null && secondNode != null) {
    if (firstNode.val < secondNode.val) {
      tail!.next = firstNode;
      firstNode = firstNode.next;
    } else {
      tail!.next = secondNode;
      secondNode = secondNode.next;
    }
    tail = tail.next;
  }

  if (firstNode != null) tail!.next = firstNode;
  if (secondNode != null) tail!.next = secondNode;

  return dummy.next!;
}
